### ISSUE TITLE
Electron crash when switching protocols

### ISSUE THREAD

**Issue by nkavian** [electron/electron#27382]
### Preflight Checklist
* [X] I have read the [Contributing Guidelines](https://github.com/electron/electron/blob/master/CONTRIBUTING.md) for this project.
* [X] I agree to follow the [Code of Conduct](https://github.com/electron/electron/blob/master/CODE_OF_CONDUCT.md) that this project adheres to.
* [X] I have searched the issue tracker for an issue that matches the one I want to file, without success.

### Issue Details

* **Electron Version:**
  * 11.2.0
* **Operating System:**
  * macOS10.14.6
* **Last Known Working Electron version:**
  * N/A

### Expected Behavior
I should be able to serve content from multiple protocols. i.e. using the standard `file://` as well as my custom protocol `foo://`.

### Actual Behavior
If I use `win.loadFile('index.html')` and stay within `file://`, there is no crash.
If I use `win.loadURL('foo://abc/deep.htm')` and stay within `foo://`, there is no crash.
If index.html has an href-link to `foo://....` then Electron crashes.

### To Reproduce

A very simple commit provided on the electron-quick-start: https://github.com/nkavian/electron-quick-start/commit/7b49ad9f2b3e98205b876c21ce3c859ea6c16877

```sh
$ git clone git@github.com:nkavian/electron-quick-start.git
$ cd electron-quick-start
$ npm install
$ npm start
```

**Comment by miniak** [comment]
```
[27527:0124/060437.761306:FATAL:navigation_request.cc(4577)] Check failed: policy->CanAccessDataForOrigin(process_id, origin). 
0   Electron Framework                  0x000000010d629e99 base::debug::CollectStackTrace(void**, unsigned long) + 9
1   Electron Framework                  0x000000010d54a2f3 base::debug::StackTrace::StackTrace() + 19
2   Electron Framework                  0x000000010d56523f logging::LogMessage::~LogMessage() + 175
3   Electron Framework                  0x000000010d5660de logging::LogMessage::~LogMessage() + 14
4   Electron Framework                  0x000000010c799e63 content::NavigationRequest::GetOriginForURLLoaderFactory() + 1139
5   Electron Framework                  0x000000010c78e911 content::NavigationRequest::CommitNavigation() + 545
6   Electron Framework                  0x000000010c798f95 content::NavigationRequest::OnWillProcessResponseChecksComplete(content::NavigationThrottle::ThrottleCheckResult) + 581
7   Electron Framework                  0x000000010c79bc7f content::NavigationRequest::OnWillProcessResponseProcessed(content::NavigationThrottle::ThrottleCheckResult) + 719
8   Electron Framework                  0x000000010c79b128 content::NavigationRequest::OnNavigationEventProcessed(content::NavigationThrottleRunner::Event, content::NavigationThrottle::ThrottleCheckResult) + 424
9   Electron Framework                  0x000000010c7a90eb content::NavigationThrottleRunner::ProcessInternal() + 1483
10  Electron Framework                  0x000000010c7a8b10 content::NavigationThrottleRunner::ProcessNavigationEvent(content::NavigationThrottleRunner::Event) + 160
11  Electron Framework                  0x000000010c7956e3 content::NavigationRequest::WillProcessResponse() + 195
12  Electron Framework                  0x000000010c795055 content::NavigationRequest::OnResponseStarted(mojo::StructPtr<network::mojom::URLLoaderClientEndpoints>, mojo::StructPtr<network::mojom::URLResponseHead>, mojo::ScopedHandleBase<mojo::DataPipeConsumerHandle>, content::GlobalRequestID, bool, content::NavigationDownloadPolicy, net::NetworkIsolationKey, base::Optional<content::SubresourceLoaderParams>) + 9077
13  Electron Framework                  0x000000010c5de3c9 content::NavigationURLLoaderImpl::NotifyResponseStarted(mojo::StructPtr<network::mojom::URLResponseHead>, mojo::StructPtr<network::mojom::URLLoaderClientEndpoints>, mojo::ScopedHandleBase<mojo::DataPipeConsumerHandle>, content::GlobalRequestID const&, bool) + 345
14  Electron Framework                  0x000000010c5e22d8 void base::internal::FunctorTraits<void (content::NavigationURLLoaderImpl::*)(mojo::StructPtr<network::mojom::URLResponseHead>, mojo::StructPtr<network::mojom::URLLoaderClientEndpoints>, mojo::ScopedHandleBase<mojo::DataPipeConsumerHandle>, content::GlobalRequestID const&, bool), void>::Invoke<void (content::NavigationURLLoaderImpl::*)(mojo::StructPtr<network::mojom::URLResponseHead>, mojo::StructPtr<network::mojom::URLLoaderClientEndpoints>, mojo::ScopedHandleBase<mojo::DataPipeConsumerHandle>, content::GlobalRequestID const&, bool), base::WeakPtr<content::NavigationURLLoaderImpl>, mojo::StructPtr<network::mojom::URLResponseHead>, mojo::StructPtr<network::mojom::URLLoaderClientEndpoints>, mojo::ScopedHandleBase<mojo::DataPipeConsumerHandle>, content::GlobalRequestID, bool>(void (content::NavigationURLLoaderImpl::*)(mojo::StructPtr<network::mojom::URLResponseHead>, mojo::StructPtr<network::mojom::URLLoaderClientEndpoints>, mojo::ScopedHandleBase<mojo::DataPipeConsumerHandle>, content::GlobalRequestID const&, bool), base::WeakPtr<content::NavigationURLLoaderImpl>&&, mojo::StructPtr<network::mojom::URLResponseHead>&&, mojo::StructPtr<network::mojom::URLLoaderClientEndpoints>&&, mojo::ScopedHandleBase<mojo::DataPipeConsumerHandle>&&, content::GlobalRequestID&&, bool&&) + 200
15  Electron Framework                  0x000000010c5e21f8 base::internal::Invoker<base::internal::BindState<void (content::NavigationURLLoaderImpl::*)(mojo::StructPtr<network::mojom::URLResponseHead>, mojo::StructPtr<network::mojom::URLLoaderClientEndpoints>, mojo::ScopedHandleBase<mojo::DataPipeConsumerHandle>, content::GlobalRequestID const&, bool), base::WeakPtr<content::NavigationURLLoaderImpl>, mojo::StructPtr<network::mojom::URLResponseHead>, mojo::StructPtr<network::mojom::URLLoaderClientEndpoints>, mojo::ScopedHandleBase<mojo::DataPipeConsumerHandle>, content::GlobalRequestID, bool>, void ()>::RunOnce(base::internal::BindStateBase*) + 72
16  Electron Framework                  0x000000010c5e2c24 _ZN4base8internal7InvokerINS0_9BindStateIZN7content23NavigationURLLoaderImpl12ParseHeadersERK4GURLPN7network5mojom15URLResponseHeadENS_12OnceCallbackIFvvEEEE3$_0JSE_SB_EEEFvN4mojo9StructPtrINS9_13ParsedHeadersEEEEE7RunOnceEPNS0_13BindStateBaseEOSK_ + 180
17  Electron Framework                  0x000000010917d745 network::mojom::NetworkContext_ParseHeaders_ForwardToCallback::Accept(mojo::Message*) + 309
18  Electron Framework                  0x000000010d9d6b98 mojo::InterfaceEndpointClient::HandleValidatedMessage(mojo::Message*) + 1608
19  Electron Framework                  0x000000010d9da42b mojo::MessageDispatcher::Accept(mojo::Message*) + 251
20  Electron Framework                  0x000000010d9d7f21 mojo::InterfaceEndpointClient::HandleIncomingMessage(mojo::Message*) + 97
21  Electron Framework                  0x000000010d9df77e mojo::internal::MultiplexRouter::ProcessIncomingMessage(mojo::internal::MultiplexRouter::MessageWrapper*, mojo::internal::MultiplexRouter::ClientCallBehavior, base::SequencedTaskRunner*) + 1278
22  Electron Framework                  0x000000010d9ded70 mojo::internal::MultiplexRouter::Accept(mojo::Message*) + 384
23  Electron Framework                  0x000000010d9da42b mojo::MessageDispatcher::Accept(mojo::Message*) + 251
24  Electron Framework                  0x000000010d9cfbbc mojo::Connector::DispatchMessage(mojo::Message) + 444
25  Electron Framework                  0x000000010d9d0797 mojo::Connector::ReadAllAvailableMessages() + 279
26  Electron Framework                  0x000000010d9d0563 mojo::Connector::OnHandleReadyInternal(unsigned int) + 99
27  Electron Framework                  0x0000000108cdd880 mojo::SimpleWatcher::DiscardReadyState(base::RepeatingCallback<void (unsigned int)> const&, unsigned int, mojo::HandleSignalsState const&) + 80
28  Electron Framework                  0x000000010d9ff218 mojo::SimpleWatcher::OnHandleReady(int, unsigned int, mojo::HandleSignalsState const&) + 440
29  Electron Framework                  0x000000010d9ff910 base::internal::Invoker<base::internal::BindState<void (mojo::SimpleWatcher::*)(int, unsigned int, mojo::HandleSignalsState const&), base::WeakPtr<mojo::SimpleWatcher>, int, unsigned int, mojo::HandleSignalsState>, void ()>::RunOnce(base::internal::BindStateBase*) + 160
30  Electron Framework                  0x000000010d5c833b base::TaskAnnotator::RunTask(char const*, base::PendingTask*) + 395
31  Electron Framework                  0x000000010d5e607f base::sequence_manager::internal::ThreadControllerWithMessagePumpImpl::DoWorkImpl(base::sequence_manager::LazyNow*) + 543
32  Electron Framework                  0x000000010d5e5cd0 base::sequence_manager::internal::ThreadControllerWithMessagePumpImpl::DoWork() + 176
33  Electron Framework                  0x000000010d640341 base::MessagePumpCFRunLoopBase::RunWork() + 65
34  Electron Framework                  0x000000010d638de2 base::mac::CallWithEHFrame(void () block_pointer) + 10
35  Electron Framework                  0x000000010d63fd4f base::MessagePumpCFRunLoopBase::RunWorkSource(void*) + 63
36  CoreFoundation                      0x00007fff38cdd884 __CFRUNLOOP_IS_CALLING_OUT_TO_A_SOURCE0_PERFORM_FUNCTION__ + 17
37  CoreFoundation                      0x00007fff38cdd823 __CFRunLoopDoSource0 + 103
38  CoreFoundation                      0x00007fff38cdd63d __CFRunLoopDoSources0 + 209
39  CoreFoundation                      0x00007fff38cdc359 __CFRunLoopRun + 937
40  CoreFoundation                      0x00007fff38cdb953 CFRunLoopRunSpecific + 466
41  HIToolbox                           0x00007fff378f6abd RunCurrentEventLoopInMode +
```

**Comment by miniak** [comment]
It crashes here https://source.chromium.org/chromium/chromium/src/+/master:content/browser/renderer_host/navigation_request.cc;l=4770?q=navigation_request.cc&ss=chromium

**Comment by miniak** [comment]
Commenting out the `CHECK` fixes the crash and the app works correctly.

**Comment by miniak** [comment]
@zcbenz or @deepak1556 can you help?

**Comment by codebytere** [comment]
To dig a little farther, the failure comes from [here](https://source.chromium.org/chromium/chromium/src/+/master:content/browser/child_process_security_policy_impl.cc;l=1684;drc=293ec036d5f0e4420241d6e4988039f14271106f)

**Comment by deepak1556** [comment]
Sorry missed this notification, this has to do with whether a process locked to a particular origin can navigate to another origin in the same process or different one safely. The above custom scheme is registered as a non standard scheme which means its origin is `null` by default.

Chromium by default doesn't allow pages loaded over file to navigate to another page that has `null` origin, try changing the link from `foo` to `data:text/html`, but in this case chromium throws a loud error instead of crash. So we can do the same to better rise the error rather than crash.

As from the user end, the custom scheme should be registered as standard scheme so that it obeys the process security policies and loads, add the following to `main.js`

```
protocol.registerSchemesAsPrivileged([
  { scheme: 'foo', privileges: { standard: true } }
])
```

**Comment by nkavian** [comment]
When I first saw the documentation for [registerSchemesAsPrivileged](https://www.electronjs.org/docs/api/protocol#protocolregisterschemesasprivilegedcustomschemes), I found it vague on certain points which lead me to believe that it might be optional altogether.

Looking today at the [privileges](https://www.electronjs.org/docs/api/structures/custom-scheme#customscheme-object), I can now appreciate I need to set some of these flags to enabled, but the documentation doesn't explain the intended behavior/consequence if a flag is on or off.

For example, if `stream` is enabled, does that mean all content has to be streamable and served via `registerStreamProtocol` or does it mean something else? Another example, performing a text search on both links above for `secure`, the word appears only once in each page, and provides no details other than stating the field is a boolean.

This ticket might be an opportunity to improve the documentation and help me and others avoid misconfiguring the `scheme`.
